API/CopilotService/__init__.py

```diff
@@ -79,26 +83,69 @@ async def _retrieve_context_from_learn_knowledge_index(context, question):
     context.custom_context.estimated_question_tokens = num_tokens_from_message(question)
     generate_results = gpt_generate(context, system_msg, question, history_msg=[])
     try:
-        task_list = _build_scenario_response(generate_results)
+        raw_task_list = _build_scenario_response(generate_results)
     except Exception as e:
         logger.error(f"Error while parsing the generate results: {generate_results}, {e}")
+        return None, None
 
     token = get_auth_token_for_learn_knowlegde_index()
 
-    splited_tasks = []
-    for task_info in task_list:
-        # use task command as task info when length of task info > 1, otherwise use task desc as task info
-        task_info = task_info.split("||")
-        task_info = task_info[1] if len(task_info) > 1 else task_info[0]
-        splited_tasks.append(asyncio.create_task(retrieve_chunks_for_atomic_task(task_info, token)))
+    context_tasks = [asyncio.create_task(_build_task_context(raw_task, token)) for raw_task in raw_task_list]
 
-    chunks_list = []
-    if len(splited_tasks) > 0:
-        chunks_list = await asyncio.gather(*splited_tasks)
+    context_info_list = await asyncio.gather(*context_tasks)
+    task_list = [context_info[0] for context_info in context_info_list]
+    chunk_list = _join_chunks_in_context(context_info_list)
 
+    chunk_list = merge_chunks_by_command(chunk_list)
     # TODO The logic of filtering, ranking, and aggregating chunks
 
-    return task_list, chunks_list
+    return task_list, chunk_list
+
+
+async def _build_task_context(raw_task, token):
+    """
+    Build a guide step and context command from a raw task
+    Args:
+        raw_task: the raw task generated by GPT(prompt for splitting task)
+    Returns: A tuple. The first element is the guide step in str and
+    the second element is a list of chunks that contain some possible commands to use.
+    """
+    desc = raw_task.split("||")[0]
+    if len(raw_task.split("||")) > 1:
+        # If GPT provides a command when splitting task, validate the command.
+        cmd = raw_task.split("||")[1]
+        validate_result = validate_command_in_task(cmd)
+        if validate_result is None:
+            # If the GPT-provided command is valid, use it as a guide step.
+            task = cmd
+            chunks = []
+        elif 'not an Azure CLI command' in validate_result.msg:
+            # If GPT generates a non-CLI command, like `git clone`
+            task = desc
+            chunks = []
+        elif 'Unknown Command' in validate_result.msg:
+            # If the GPT-provided command has an error in command signature,
+            # retrieve context chunks according to the description
+            task = desc
```

Comment:
```suggestion
            # If the GPT-provided command has an error in command signature, 
            # retrieve context chunks according to the command signature
            # because hallucination command signatures preserve the semantics of operations without attention issues,
            # and there are always only subtle differences between the hallucination and correct command signatures.
```